x/crosschain/keeper/grpc_query_cctx_rate_limit_test.go

```diff
@@ -123,60 +124,486 @@ func setupForeignCoins(
+			ethMindedCctxs:  ethMindedCctxs,
+			ethPendingCctxs: ethPendingCctxs,
+			ethPendingNonces: observertypes.PendingNonces{
+				ChainId:   ethChainID,
+				NonceLow:  1099,
+				NonceHigh: 1199,
+				Tss:       tss.TssPubkey,
+			},
+			btcMinedCctxs:   btcMinedCctxs,
+			btcPendingCctxs: btcPendingCctxs,
+			btcPendingNonces: observertypes.PendingNonces{
+				ChainId:   btcChainID,
+				NonceLow:  1099,
+				NonceHigh: 1199,
+				Tss:       tss.TssPubkey,
+			},
+			currentHeight:        1199,
+			queryLimit:           keeper.MaxPendingCctxs,
+			expectedCctxs:        append(append([]*types.CrossChainTx{}, ethPendingCctxs...), btcPendingCctxs...),
+			expectedTotalPending: 400,
+			expectedTotalValue:   1500, // 500 (window) * (2.5 + 0.5)
+			rateLimitExceeded:    false,
+		},
+		{
+			name:            "can loop backwards all the way to endNonce 0",
+			rateLimitFlags:  createTestRateLimiterFlags(500, math.NewUint(5000), zrc20ETH, zrc20BTC, zrc20USDT, "2500", "50000", "0.8"),
+			ethMindedCctxs:  ethMindedCctxs,
+			ethPendingCctxs: ethPendingCctxs,
+			ethPendingNonces: observertypes.PendingNonces{
+				ChainId:   ethChainID,
+				NonceLow:  999, // endNonce will set to 0 as NonceLow - 1000 < 0
+				NonceHigh: 1199,
+				Tss:       tss.TssPubkey,
+			},
+			btcMinedCctxs:   btcMinedCctxs,
+			btcPendingCctxs: btcPendingCctxs,
+			btcPendingNonces: observertypes.PendingNonces{
+				ChainId:   btcChainID,
+				NonceLow:  999, // endNonce will set to 0 as NonceLow - 1000 < 0
+				NonceHigh: 1199,
+				Tss:       tss.TssPubkey,
+			},
+			currentHeight:        1199,
+			queryLimit:           keeper.MaxPendingCctxs,
+			expectedCctxs:        append(append([]*types.CrossChainTx{}, ethPendingCctxs...), btcPendingCctxs...),
+			expectedTotalPending: 400,
+			expectedTotalValue:   1500, // 500 (window) * (2.5 + 0.5)
+			rateLimitExceeded:    false,
+		},
+		{
+			name:            "set a low rate (< 1200) to early break the LoopBackwards with criteria #2",
+			rateLimitFlags:  createTestRateLimiterFlags(500, math.NewUint(1000), zrc20ETH, zrc20BTC, zrc20USDT, "2500", "50000", "0.8"), // 1000 < 1200
+			ethMindedCctxs:  ethMindedCctxs,
+			ethPendingCctxs: ethPendingCctxs,
+			ethPendingNonces: observertypes.PendingNonces{
+				ChainId:   ethChainID,
+				NonceLow:  1099,
+				NonceHigh: 1199,
+				Tss:       tss.TssPubkey,
+			},
+			btcMinedCctxs:   btcMinedCctxs,
+			btcPendingCctxs: btcPendingCctxs,
+			btcPendingNonces: observertypes.PendingNonces{
+				ChainId:   btcChainID,
+				NonceLow:  1099,
+				NonceHigh: 1199,
+				Tss:       tss.TssPubkey,
+			},
+			currentHeight:        1199,
+			queryLimit:           keeper.MaxPendingCctxs,
+			expectedCctxs:        append(append([]*types.CrossChainTx{}, ethPendingCctxs[0:100]...), btcPendingCctxs[0:100]...),
+			expectedTotalPending: 400,
+			rateLimitExceeded:    true,
+		},
+		{
+			name: "set high rate and wide window to early to break inner loop with the criteria #1",
```

Comment:
what is criteria 1?

Reply:
I'll add more comment to this. It's intended for this break: https://github.com/zeta-chain/node/blob/unit-test-rate-limit/x/crosschain/keeper/grpc_query_cctx_rate_limit.go#L173

Reply:
`set high rate and wide window to break inner loop of LoopBackwards when left window boundary is reached`

Reply:
https://github.com/zeta-chain/node/pull/2060/commits/578afccb54c86839c9c505dab05893132164b7ee